### ISSUE TITLE
Powerset generator in EFS consumed 240 GB of RAM for ~ 300 features

### ISSUE THREAD

**Issue by adam-erickson** [rasbt/mlxtend#194]
The _candidates_ object in _EFS.fit_ implements a standard powerset generator using _itertools_. This function quickly consumed 240 GB of RAM with ~300 features for the range (0, 20). This makes sense given that this generates around 7.5^30 8-byte integers, making it too large for memory even in a cluster environment.

I'm currently testing a solution utilizing _joblib_ _delayed_ and _Parallel_ along with _dask.array_, which looks like this:

```
import dask.array as da

def powerset(iterable, min_features, max_features)
    """
    returns a generator following the form:
    powerset([1,2,3] for range [1,3]) --> (1,) (2,) (3,) (1,2) (1,3) (2,3) (1,2,3)
    """
    s = list(iterable)
    return chain.from_iterable(combinations(s, r) for r in range(min_features, max_features + 1))

min_features, max_features = [1, 20]
ps = powerset(iterable=X, min_features=min_features, max_features)
def ret(x):
    return x
ps = [Parallel(n_jobs=4)(delayed(ret)(i) for i in ps)]
ps = da.from_delayed(ps)
```

So far, memory consumption appears dramatically reduced, but CPU utilization is 100% on a single core.

Cheers,

Adam

**Comment by rasbt** [comment]
Thanks for the feedback. I never thought someone would actually run this exhaustive feature selection on such a large set of features, but yeah, why not fixing it so that it at least theoretically works :).

So, instead of generating the candidates upfront, we could use sth similar to your solution above (which is really elegant btw!) or a simple generator like so:

```python
from itertools import combinations


def powerset(min_features, max_features):
    for ln in range(min_features, max_features+1):
        for feature_set in combinations(range(max_features), r=ln):
            if feature_set:
                yield feature_set
    
            
min_features = 3
max_features = 8

for candidate in powerset(min_features, max_features):
    # evaluate
    pass
```

I can take a look at, or if you'd like to submit a PR, that'd be nice :)

**Comment by adam-erickson** [comment]
PR added. The trouble with the powerset function you posted above is that it doesn't take into account the number of possible items to draw from, but is fixed as equal to max_features. The function I originally posted above would have required a list of column names or indices, which is also slightly less efficient.

**Comment by rasbt** [comment]
Fixed via #195
